### PR TITLE
varlink: add ability to talk to remote Varlink service binary via ssh

### PR DESCRIPTION
**Pull request by poettering**
So far our varlink client could connect to remote AF_UNIX sockets via SSH. This adds a 2nd SSH-based transport: invoke a binary remotely via ssh, and talk to it via stdin/stdout.

This is v257 material, don't bother with this for v256

### CODE REVIEW

File: src/shared/varlink.c [hunk 1]
```diff
@@ -1691,22 +1700,43 @@ int varlink_wait(Varlink *v, usec_t timeout) {
...
-        r = fd_wait_for_event(fd, events, t);
+        struct pollfd pollfd[2];
+        size_t n_poll_fd = 0;
+
+        if (v->input_fd == v->output_fd) {
+                pollfd[n_poll_fd++] = (struct pollfd) {
+                        .fd = v->input_fd,
+                        .events = events,
+                };
+
```
Comment (daandemeyer) [inline src/shared/varlink.c]: Spurious newline

File: man/varlinkctl.xml [hunk 2]
```diff
@@ -71,16 +71,23 @@
...
 
-      <listitem><para>A Varlink service reference starting with the <literal>ssh:</literal> string, followed
+      <listitem><para>A Varlink service reference starting with the <literal>ssh-unix:</literal> string, followed
       by an SSH host specification, followed by <literal>:</literal>, followed by an absolute
       <constant>AF_UNIX</constant> socket path. (This requires OpenSSH 9.4 or newer on the server side,
       abstract namespace sockets are not supported.)</para></listitem>
+
+      <listitem><para>A Varlink service reference starting with the <literal>ssh-exec:</literal> string,
+      followed by an SSH host specification, followed by <literal>:</literal>, followed by a command line. In
+      this case the comman is invoked and the Varlink protocol spoken on the standard input and output of the
```
Comment (daandemeyer) [inline man/varlinkctl.xml]: comman => command

Comment (daandemeyer) [reply man/varlinkctl.xml]: "Varlink protocol is spoken"

File: src/shared/varlink.c [hunk 3]
```diff
@@ -596,13 +609,107 @@ static int varlink_connect_ssh(Varlink **ret, const char *where) {
...
+        const char *e = strchr(where, ':');
+        if (!e)
+                return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "SSH specification lacks a : separator between host and path, refusing: %s", where);
+
+        _cleanup_free_ char *h = strndup(where, e - where);
+        if (!h)
+                return log_oom_debug();
+
+        _cleanup_strv_free_ char **cmdline = NULL;
+        r = strv_split_full(&cmdline, e + 1, /* serparators= */ NULL, EXTRACT_CUNESCAPE|EXTRACT_UNQUOTE);
```
Comment (daandemeyer) [inline src/shared/varlink.c]: "separators"

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]
This could really do with some kind of integration test.

A man page example would also help a lot in understanding how to use this.

Also, the title feels wrong in a way. This allows you to invoke a command on a remote host  and communicate with it using varlink. On its own it doesn't really allow you to talk to a running service on a remote host using varlink.

Shouldn't this also add systemd-varlink-bridge like we have systemd-stdio-bridge for D-Bus which connects to a varlink service running on a remote host?

**Comment by poettering** [comment]
> Shouldn't this also add systemd-varlink-bridge like we have systemd-stdio-bridge for D-Bus which connects to a varlink service running on a remote host?

There's no need, we already support that. See varlinkctl(1) man page. You can already access remote services that are bound to some AF_UNIX support. This requires the new support for connecting to a remote AF_UNIX socket via "ssh -W". Only available since OpenSSH 9.4, which is really recent though.

systemd-stdio-bridge is a concept from a time where "ssh -W" didn't exist. I mean, we can't kill it for now, as older ssh versions are still popular. And maybe it even makes sense to keep after that, since it actually patches the negotation protocol of D-Bus to disable fd passing, since that doesn't work via ssh. (This concept like this doesn't exist in varlink anyway, hence no need for anything like this in a varlink world).

To summarize, with this PR we support two different things:

1. varlinkctl ssh-unix:hostxyz:/run/some/socket
2. varlinkctl ssh-exec:hostabc:/usr/bin/some/binary

And neither requires any particular tool on the other side to invoke, we just need the target socket or binary to talk to.

**Comment by poettering** [comment]
rebased, addressed the issues, added examples to the varlinkctl man page and a CI test. ptal.

**Review by DaanDeMeyer — APPROVED** [review]
